Perform profiling of the resource consumption and consider optimsations

## Observation

The memory usage goes up upon repetitive upload on a file. 

Opens questions:

- Does the level make sense? 
- Can it be optimized, and by how much? 
- Does the observation approach make sense?
- What would go profiler show?

## How to reproduce

1. Clone the repo
2. Initialise local demo setup:

```commandline
docker compose up --build
```

3. Open the container usage statistics:

```commandline
docker stats $(docker ps --filter "label=com.docker.compose.service=gateway-container" --format "{{.ID}}")
```

4. Upload a text:

```commanline
echo "foo bar baz" > tinytext.txt && curl -T tinytext.txt http://localhost:3000/object/tinytext || rm tinytext.txt && if [ -f tinytext.txt ]; then rm tinytext.txt; fi
```

Observe memory usage going up.

## Illustration

### Data

|Request|RAM consumption [MiB]|
|:---|:---|
|0|2|
|1|12|
|2|22|
|3|551|
|4|1056|
|5|1056|
|6|1056|
|7|1056|
|8|1056|
|9|1056|
|10|1056|

### Initial state - no requests
<img width="931" alt="image" src="https://github.com/kislerdm/object-storage-gateway/assets/13434797/fbdd2d82-576e-4084-a510-8dbf7d23b9d2">

### After the 1st PUT request

<img width="973" alt="image" src="https://github.com/kislerdm/object-storage-gateway/assets/13434797/2fa94459-2c96-494e-af34-00a4f5f7ece1">

### After the 2nd PUT requests

<img width="963" alt="image" src="https://github.com/kislerdm/object-storage-gateway/assets/13434797/7e59d4bc-e788-428b-989f-a67a285e2e7f">

### After the 3rd PUT request

<img width="978" alt="image" src="https://github.com/kislerdm/object-storage-gateway/assets/13434797/73615268-3110-4041-b569-b4cf4b0a4b4a">

### After the 4rd PUT request

<img width="989" alt="image" src="https://github.com/kislerdm/object-storage-gateway/assets/13434797/d05e942b-8518-4545-8bb9-df7f584ffab4">

### After the 5th PUT request

<img width="982" alt="image" src="https://github.com/kislerdm/object-storage-gateway/assets/13434797/b57636a4-ab55-4fd9-9bb2-571671ed8e98">

### After the 10th PUT request

<img width="956" alt="image" src="https://github.com/kislerdm/object-storage-gateway/assets/13434797/62c63dc1-4bab-4d37-9c11-951477c98220">

The gateway fails consuming all available resources even though the container is allocated 11Gb RAM limit.

```commandline
{"time":"2023-10-11T01:57:26.527406+02:00","level":"INFO","msg":"comparison","objectID":"16","match":true}
{"time":"2023-10-11T01:57:27.009497+02:00","level":"INFO","msg":"comparison","objectID":"21","match":true}
{"time":"2023-10-11T01:57:27.121382+02:00","level":"INFO","msg":"comparison","objectID":"86","match":true}
{"time":"2023-10-11T01:57:27.313196+02:00","level":"INFO","msg":"comparison","objectID":"43","match":true}
{"time":"2023-10-11T01:57:27.313375+02:00","level":"INFO","msg":"comparison","objectID":"30","match":true}
{"time":"2023-10-11T01:57:27.386124+02:00","level":"INFO","msg":"comparison","objectID":"6","match":true}
{"time":"2023-10-11T01:57:27.567282+02:00","level":"INFO","msg":"comparison","objectID":"93","match":true}
{"time":"2023-10-11T01:57:27.670712+02:00","level":"INFO","msg":"comparison","objectID":"74","match":true}
{"time":"2023-10-11T01:57:27.696865+02:00","level":"INFO","msg":"comparison","objectID":"83","match":true}
{"time":"2023-10-11T01:57:27.69955+02:00","level":"INFO","msg":"comparison","objectID":"80","match":true}
{"time":"2023-10-11T01:57:27.700247+02:00","level":"INFO","msg":"comparison","objectID":"95","match":true}
{"time":"2023-10-11T01:57:27.798759+02:00","level":"INFO","msg":"comparison","objectID":"37","match":true}
{"time":"2023-10-11T01:57:27.881364+02:00","level":"INFO","msg":"comparison","objectID":"42","match":true}
{"time":"2023-10-11T01:57:27.898783+02:00","level":"INFO","msg":"comparison","objectID":"96","match":true}
{"time":"2023-10-11T01:57:27.901894+02:00","level":"INFO","msg":"comparison","objectID":"19","match":true}
{"time":"2023-10-11T01:57:27.902128+02:00","level":"INFO","msg":"comparison","objectID":"15","match":true}
{"time":"2023-10-11T01:57:27.902233+02:00","level":"INFO","msg":"comparison","objectID":"58","match":true}
{"time":"2023-10-11T01:57:28.268609+02:00","level":"INFO","msg":"comparison","objectID":"76","match":true}
{"time":"2023-10-11T01:57:28.269788+02:00","level":"INFO","msg":"comparison","objectID":"85","match":true}
{"time":"2023-10-11T01:57:28.26992+02:00","level":"INFO","msg":"comparison","objectID":"90","match":true}
{"time":"2023-10-11T01:57:28.397856+02:00","level":"INFO","msg":"comparison","objectID":"63","match":true}
{"time":"2023-10-11T01:57:28.481482+02:00","level":"INFO","msg":"comparison","objectID":"2","match":true}
{"time":"2023-10-11T01:57:28.491465+02:00","level":"INFO","msg":"comparison","objectID":"75","match":true}
{"time":"2023-10-11T01:57:28.523376+02:00","level":"INFO","msg":"comparison","objectID":"3","match":true}
{"time":"2023-10-11T01:57:28.643083+02:00","level":"INFO","msg":"comparison","objectID":"39","match":true}
{"time":"2023-10-11T01:57:28.693573+02:00","level":"INFO","msg":"comparison","objectID":"27","match":true}
{"time":"2023-10-11T01:57:28.726053+02:00","level":"INFO","msg":"comparison","objectID":"48","match":true}
{"time":"2023-10-11T01:57:28.727154+02:00","level":"INFO","msg":"comparison","objectID":"84","match":true}
{"time":"2023-10-11T01:57:28.727805+02:00","level":"INFO","msg":"comparison","objectID":"98","match":true}
{"time":"2023-10-11T01:57:28.727908+02:00","level":"INFO","msg":"comparison","objectID":"51","match":true}
{"time":"2023-10-11T01:57:28.771055+02:00","level":"INFO","msg":"comparison","objectID":"57","match":true}
{"time":"2023-10-11T01:57:28.809615+02:00","level":"INFO","msg":"comparison","objectID":"18","match":true}
{"time":"2023-10-11T01:57:28.813475+02:00","level":"INFO","msg":"comparison","objectID":"53","match":true}
{"time":"2023-10-11T01:57:28.848936+02:00","level":"INFO","msg":"comparison","objectID":"54","match":true}
{"time":"2023-10-11T01:57:28.850682+02:00","level":"INFO","msg":"comparison","objectID":"55","match":true}
{"time":"2023-10-11T01:57:28.860007+02:00","level":"INFO","msg":"comparison","objectID":"12","match":true}
{"time":"2023-10-11T01:57:28.952385+02:00","level":"INFO","msg":"comparison","objectID":"7","match":true}
{"time":"2023-10-11T01:57:28.953118+02:00","level":"INFO","msg":"comparison","objectID":"25","match":true}
{"time":"2023-10-11T01:57:28.993075+02:00","level":"INFO","msg":"comparison","objectID":"38","match":true}
{"time":"2023-10-11T01:57:29.023588+02:00","level":"INFO","msg":"comparison","objectID":"28","match":true}
{"time":"2023-10-11T01:57:29.044709+02:00","level":"INFO","msg":"comparison","objectID":"4","match":true}
{"time":"2023-10-11T01:57:29.06512+02:00","level":"INFO","msg":"comparison","objectID":"52","match":true}
{"time":"2023-10-11T01:57:29.097901+02:00","level":"INFO","msg":"comparison","objectID":"81","match":true}
{"time":"2023-10-11T01:57:29.098876+02:00","level":"INFO","msg":"comparison","objectID":"82","match":true}
{"time":"2023-10-11T01:57:29.221423+02:00","level":"INFO","msg":"comparison","objectID":"31","match":true}
{"time":"2023-10-11T01:57:29.221514+02:00","level":"INFO","msg":"comparison","objectID":"70","match":true}
{"time":"2023-10-11T01:57:29.221886+02:00","level":"INFO","msg":"comparison","objectID":"67","match":true}
{"time":"2023-10-11T01:57:29.22199+02:00","level":"INFO","msg":"comparison","objectID":"91","match":true}
{"time":"2023-10-11T01:57:29.231556+02:00","level":"INFO","msg":"comparison","objectID":"22","match":true}
{"time":"2023-10-11T01:57:29.231769+02:00","level":"INFO","msg":"comparison","objectID":"71","match":true}
{"time":"2023-10-11T01:57:29.232395+02:00","level":"INFO","msg":"comparison","objectID":"65","match":true}
{"time":"2023-10-11T01:57:29.309556+02:00","level":"INFO","msg":"comparison","objectID":"59","match":true}
{"time":"2023-10-11T01:57:29.309683+02:00","level":"INFO","msg":"comparison","objectID":"35","match":true}
{"time":"2023-10-11T01:57:29.326346+02:00","level":"INFO","msg":"comparison","objectID":"44","match":true}
{"time":"2023-10-11T01:57:29.355453+02:00","level":"INFO","msg":"comparison","objectID":"62","match":true}
{"time":"2023-10-11T01:57:29.361857+02:00","level":"INFO","msg":"comparison","objectID":"5","match":true}
{"time":"2023-10-11T01:57:29.454702+02:00","level":"INFO","msg":"comparison","objectID":"29","match":true}
{"time":"2023-10-11T01:57:29.548096+02:00","level":"INFO","msg":"comparison","objectID":"88","match":true}
{"time":"2023-10-11T01:57:29.619761+02:00","level":"INFO","msg":"comparison","objectID":"56","match":true}
{"time":"2023-10-11T01:57:29.620254+02:00","level":"INFO","msg":"comparison","objectID":"41","match":true}
{"time":"2023-10-11T01:57:43.539618+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/1\": EOF","step":"write","objectID":"1"}
{"time":"2023-10-11T01:57:43.539619+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/73\": EOF","step":"write","objectID":"73"}
{"time":"2023-10-11T01:57:43.540428+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/46\": EOF","step":"write","objectID":"46"}
{"time":"2023-10-11T01:57:43.540436+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/23\": EOF","step":"write","objectID":"23"}
{"time":"2023-10-11T01:57:43.540453+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/49\": EOF","step":"write","objectID":"49"}
{"time":"2023-10-11T01:57:43.540577+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/8\": EOF","step":"write","objectID":"8"}
{"time":"2023-10-11T01:57:43.540641+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/40\": EOF","step":"write","objectID":"40"}
{"time":"2023-10-11T01:57:43.540774+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/14\": EOF","step":"write","objectID":"14"}
{"time":"2023-10-11T01:57:43.540829+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/89\": EOF","step":"write","objectID":"89"}
{"time":"2023-10-11T01:57:43.540847+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/45\": EOF","step":"write","objectID":"45"}
{"time":"2023-10-11T01:57:43.540871+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/50\": EOF","step":"write","objectID":"50"}
{"time":"2023-10-11T01:57:43.540888+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/77\": EOF","step":"write","objectID":"77"}
{"time":"2023-10-11T01:57:43.540897+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/9\": EOF","step":"write","objectID":"9"}
{"time":"2023-10-11T01:57:43.540918+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/47\": EOF","step":"write","objectID":"47"}
{"time":"2023-10-11T01:57:43.540923+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/79\": EOF","step":"write","objectID":"79"}
{"time":"2023-10-11T01:57:43.540927+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/24\": EOF","step":"write","objectID":"24"}
{"time":"2023-10-11T01:57:43.540949+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/26\": EOF","step":"write","objectID":"26"}
{"time":"2023-10-11T01:57:43.540962+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/20\": EOF","step":"write","objectID":"20"}
{"time":"2023-10-11T01:57:43.541358+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/60\": EOF","step":"write","objectID":"60"}
{"time":"2023-10-11T01:57:43.541429+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/64\": EOF","step":"write","objectID":"64"}
{"time":"2023-10-11T01:57:43.54158+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/69\": EOF","step":"write","objectID":"69"}
{"time":"2023-10-11T01:57:43.541603+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/33\": EOF","step":"write","objectID":"33"}
{"time":"2023-10-11T01:57:43.541734+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/36\": EOF","step":"write","objectID":"36"}
{"time":"2023-10-11T01:57:43.541736+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/11\": EOF","step":"write","objectID":"11"}
{"time":"2023-10-11T01:57:43.541796+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/66\": EOF","step":"write","objectID":"66"}
{"time":"2023-10-11T01:57:43.541826+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/99\": EOF","step":"write","objectID":"99"}
{"time":"2023-10-11T01:57:43.541893+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/34\": EOF","step":"write","objectID":"34"}
{"time":"2023-10-11T01:57:43.541931+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/78\": EOF","step":"write","objectID":"78"}
{"time":"2023-10-11T01:57:43.54194+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/94\": EOF","step":"write","objectID":"94"}
{"time":"2023-10-11T01:57:43.54196+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/72\": EOF","step":"write","objectID":"72"}
{"time":"2023-10-11T01:57:43.54206+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/92\": EOF","step":"write","objectID":"92"}
{"time":"2023-10-11T01:57:43.542064+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/17\": EOF","step":"write","objectID":"17"}
{"time":"2023-10-11T01:57:43.542069+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/68\": EOF","step":"write","objectID":"68"}
{"time":"2023-10-11T01:57:43.542074+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/61\": EOF","step":"write","objectID":"61"}
{"time":"2023-10-11T01:57:43.542078+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/97\": EOF","step":"write","objectID":"97"}
{"time":"2023-10-11T01:57:43.542206+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/32\": EOF","step":"write","objectID":"32"}
{"time":"2023-10-11T01:57:43.542213+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/13\": EOF","step":"write","objectID":"13"}
{"time":"2023-10-11T01:57:43.542233+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/87\": EOF","step":"write","objectID":"87"}
{"time":"2023-10-11T01:57:43.542242+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/0\": EOF","step":"write","objectID":"0"}
{"time":"2023-10-11T01:57:43.543801+02:00","level":"ERROR","msg":"Put \"http://localhost:3000/object/10\": EOF","step":"write","objectID":"10"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x1047078f0]

goroutine 61 [running]:
main.read(0x14000110110?, {0x10471c5c6?, 0x1049c0780?})
	/Users/dkisler/projects/spacelift/homework-object-storage/scripts/loadtest/main.go:54 +0x130
main.main.func1(0x0?, 0x140001200f0)
	/Users/dkisler/projects/spacelift/homework-object-storage/scripts/loadtest/main.go:100 +0x1b4
created by main.main in goroutine 1
	/Users/dkisler/projects/spacelift/homework-object-storage/scripts/loadtest/main.go:91 +0x204
exit status 2
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,4 +1,15 @@
-## Unreleased
+## v0.0.6
+
+### Added
+
+- [BREAKING] [[#2](https://github.com/kislerdm/object-storage-gateway/issues/2)] The configuration `objectSizeBytes` was added as the 
+attribute of the `Gateway`'s `Write` method. The change enables to reduce memory consumption significantly: 
+execution of the use case described in the [issue](https://github.com/kislerdm/object-storage-gateway/issues/2) will require ~10MiB of RAM instead of ~1GiB.
+
+---
+**Note** that `objectSizeBytes` can be set to `-1` when the object size is unknown. 
+However, it will result in a greedy memory allocation strategy identical to the one used in the previous releases.
+---
 
 ### v0.0.5
 

--- a/README.md
+++ b/README.md
@@ -109,7 +109,7 @@ classDiagram
        +Logger *slog.Logger
 
       +Read(ctx context.Context, id string) io.ReadCloser, bool, error
-      +Write(ctx context.Context, id string, reader io.Reader) error
+      +Write(ctx context.Context, id string, reader io.Reader, objectSizeBytes int64) error
   }
 
   class StorageConnectionReadFinder {
@@ -123,7 +123,7 @@ classDiagram
       // pkg/gateway/gateway.go
       <<interface>>
       Read(ctx context.Context, bucketName, objectName string) io.ReadCloser, bool, error
-      Write(ctx context.Context, bucketName, objectName string, reader io.Reader) error
+      Write(ctx context.Context, bucketName, objectName string, reader io.Reader, objectSizeBytes int64) error
       Find(ctx context.Context, bucketName, objectName string) bool, error
   }
 

--- a/internal/minio/minio.go
+++ b/internal/minio/minio.go
@@ -46,7 +46,7 @@ func (c *Client) Read(ctx context.Context, bucketName, objectName string) (io.Re
 	return reader, true, nil
 }
 
-func (c *Client) Write(ctx context.Context, bucketName, objectName string, reader io.Reader) error {
+func (c *Client) Write(ctx context.Context, bucketName, objectName string, reader io.Reader, objectSizeBytes int64) error {
 	exists, err := c.BucketExists(ctx, bucketName)
 	if err != nil {
 		return fmt.Errorf("cannot store the object: %w", err)
@@ -57,7 +57,7 @@ func (c *Client) Write(ctx context.Context, bucketName, objectName string, reade
 		}
 	}
 
-	_, err = c.PutObject(ctx, bucketName, objectName, reader, -1, minio.PutObjectOptions{})
+	_, err = c.PutObject(ctx, bucketName, objectName, reader, objectSizeBytes, minio.PutObjectOptions{})
 	return err
 }
 

--- a/internal/restfulhandler/handler.go
+++ b/internal/restfulhandler/handler.go
@@ -6,6 +6,7 @@ import (
 	"log/slog"
 	"net/http"
 	"os"
+	"strconv"
 	"strings"
 
 	"github.com/kislerdm/object-storage-gateway/pkg/gateway"
@@ -94,7 +95,7 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		}
 
 		defer func() { _ = r.Body.Close() }()
-		if err := h.rw.Write(r.Context(), objectID, r.Body); err != nil {
+		if err := h.rw.Write(r.Context(), objectID, r.Body, contentSize(r)); err != nil {
 			h.logError(r, http.StatusInternalServerError, err.Error())
 			writeErrorMessage(w, http.StatusInternalServerError, "failed to write object")
 			return
@@ -108,6 +109,14 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		writeErrorMessage(w, http.StatusMethodNotAllowed, "method not allowed")
 		return
 	}
+}
+
+func contentSize(r *http.Request) int64 {
+	v, err := strconv.ParseInt(r.Header.Get("Content-Length"), 10, 64)
+	if err != nil {
+		return -1
+	}
+	return v
 }
 
 func concatHeaders(headers http.Header) string {
@@ -167,5 +176,5 @@ func writeErrorMessage(w http.ResponseWriter, statusCode int, s string) {
 // reader defines the interface to store and retrieve data.
 type readWriter interface {
 	Read(ctx context.Context, id string) (readCloser io.ReadCloser, found bool, err error)
-	Write(ctx context.Context, id string, reader io.Reader) error
+	Write(ctx context.Context, id string, reader io.Reader, objectSizeBytes int64) error
 }

--- a/internal/restfulhandler/handler_test.go
+++ b/internal/restfulhandler/handler_test.go
@@ -42,7 +42,7 @@ func (m *mockReadWriter) Read(_ context.Context, _ string) (readCloser io.ReadCl
 	return io.NopCloser(m.readCloser), m.readCloser != nil, nil
 }
 
-func (m *mockReadWriter) Write(_ context.Context, _ string, reader io.Reader) error {
+func (m *mockReadWriter) Write(_ context.Context, _ string, reader io.Reader, _ int64) error {
 	if m.err != nil {
 		return m.err
 	}

--- a/pkg/gateway/gateway.go
+++ b/pkg/gateway/gateway.go
@@ -115,7 +115,7 @@ func (s *Gateway) Read(ctx context.Context, id string) (io.ReadCloser, bool, err
 }
 
 // Write writes object to the storage.
-func (s *Gateway) Write(ctx context.Context, id string, reader io.Reader) error {
+func (s *Gateway) Write(ctx context.Context, id string, reader io.Reader, objectSizeBytes int64) error {
 	instances, err := s.storageDiscoveryClient.Find(ctx, s.storageInstancesSelector)
 	if err != nil {
 		return err
@@ -153,7 +153,7 @@ func (s *Gateway) Write(ctx context.Context, id string, reader io.Reader) error 
 				slog.String("objectID", id),
 			)
 
-			return conn.Write(ctx, s.storageBucket, id, reader)
+			return conn.Write(ctx, s.storageBucket, id, reader, objectSizeBytes)
 		}
 	}
 
@@ -171,7 +171,7 @@ func (s *Gateway) Write(ctx context.Context, id string, reader io.Reader) error 
 		slog.String("objectID", id),
 	)
 
-	return conn.Write(ctx, s.storageBucket, id, reader)
+	return conn.Write(ctx, s.storageBucket, id, reader, objectSizeBytes)
 }
 
 func (s *Gateway) newStorageInstanceConnection(ctx context.Context, id string) (ObjectReadWriteFinder, error) {
@@ -233,7 +233,7 @@ type ObjectReadWriteFinder interface {
 	Read(ctx context.Context, bucketName, objectName string) (io.ReadCloser, bool, error)
 
 	// Write writes the object.
-	Write(ctx context.Context, bucketName, objectName string, reader io.Reader) error
+	Write(ctx context.Context, bucketName, objectName string, reader io.Reader, objectSizeBytes int64) error
 
 	// Find identifies if the object can be found in the instance.
 	Find(ctx context.Context, bucketName, objectName string) (bool, error)

--- a/pkg/gateway/gateway_test.go
+++ b/pkg/gateway/gateway_test.go
@@ -103,7 +103,7 @@ func TestGateway_Write(t *testing.T) {
 		gateway.newStorageConnectionFn = mockMinioConnectionFactory(nil, &mockStorageClient{dataReader: inputData})
 
 		// WHEN
-		err := gateway.Write(context.TODO(), inputID, inputData)
+		err := gateway.Write(context.TODO(), inputID, inputData, -1)
 
 		// THEN
 		if err != nil {
@@ -117,7 +117,7 @@ func TestGateway_Write(t *testing.T) {
 		gateway.newStorageConnectionFn = mockMinioConnectionFactory(nil, &mockStorageClient{})
 
 		// WHEN
-		err := gateway.Write(context.TODO(), inputID, inputData)
+		err := gateway.Write(context.TODO(), inputID, inputData, -1)
 
 		// THEN
 		if err != nil {
@@ -148,7 +148,7 @@ func (m *mockStorageClient) Read(_ context.Context, _, _ string) (io.ReadCloser,
 	return io.NopCloser(m.dataReader), m.dataReader != nil, nil
 }
 
-func (m *mockStorageClient) Write(_ context.Context, _, _ string, reader io.Reader) error {
+func (m *mockStorageClient) Write(_ context.Context, _, _ string, reader io.Reader, _ int64) error {
 	if m.err != nil {
 		return m.err
 	}

